server/src/main/java/org/opensearch/tasks/TaskCancellationMonitoringService.java

```diff
@@ -0,0 +1,202 @@
+                    return true;
+                } else {
+                    return false;
+                }
+            }).count();
+            cancellationStatsHolder.get(key).totalLongRunningCancelledTaskCount.inc(uniqueTasksRunningCount);
+        });
+    }
+
+    @Override
+    protected void doStart() {
+        scheduledFuture = threadPool.scheduleWithFixedDelay(() -> {
+            try {
+                doRun();
+            } catch (Exception e) {
+                logger.debug("Exception occurred in Task monitoring service", e);
+            }
+        }, taskCancellationMonitoringSettings.getInterval(), ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    protected void doStop() {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel();
+        }
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+
+    }
+
+    // For testing
+    protected Map<Long, Boolean> getCancelledTaskTracker() {
+        return this.cancelledTaskTracker;
+    }
+
+    // For testing
+    protected boolean shouldRun() {
+        return shouldRun.get();
+    }
+
+    /**
+     * Invoked when a task is completed. This helps us to disable monitoring service when there are no cancelled tasks
+     * running to avoid wasteful work.
+     * @param task task which got completed.
+     */
+    @Override
+    public void onTaskCompleted(Task task) {
+        if (!TASKS_TO_TRACK.contains(task.getClass())) {
+            return;
+        }
+        if (!this.cancelledTaskTracker.containsKey(task.getId())) {
+            return;
+        }
+        this.cancelledTaskTracker.remove(task.getId());
+        if (this.cancelledTaskTracker.isEmpty()) {
+            shouldRun.set(false);
+        }
+    }
+
+    /**
+     * Invoked when a task is cancelled. This is to keep track of tasks being cancelled. More importantly also helps
+     * us to enable this monitoring service only when needed.
+     * @param task task which got cancelled.
+     */
+    @Override
+    public void onTaskCancelled(CancellableTask task) {
+        if (!TASKS_TO_TRACK.contains(task.getClass())) {
+            return;
+        }
+        // Add task to tracker and mark it as not seen(false) yet by the stats logic.
+        this.cancelledTaskTracker.putIfAbsent(task.getId(), false);
+        if (!shouldRun.get()) {
+            shouldRun.set(true);
+        }
+    }
+
+    public TaskCancellationStats stats() {
+        Map<Class<? extends CancellableTask>, List<CancellableTask>> currentRunningCancelledTasks =
+            getCurrentRunningTasksPostCancellation();
+        return new TaskCancellationStats(
+            new SearchShardTaskCancellationStats(
+                Optional.of(currentRunningCancelledTasks).map(mapper -> mapper.get(SearchShardTask.class)).map(List::size).orElse(0),
+                cancellationStatsHolder.get(SearchShardTask.class).totalLongRunningCancelledTaskCount.count()
+            )
```

Comment:
Why not SearchTask too? Should this be a map instead of task type to detailed stats per type

Reply:
For now, we want to track SearchShardTask. Will add this later if needed.

Reply:
Also created an issue to track this - https://github.com/opensearch-project/OpenSearch/issues/7977